posix: implement other `_POSIX_THREAD_SAFE_FUNCTIONS`

ycsin opened the pull request:
> [!NOTE]
> Has a commit from https://github.com/zephyrproject-rtos/zephyr/pull/74629

Time-related functions are in #74180.

This PR implements:
- `flockfile()`
- `ftrylockfile()`
- `funlockfile()`
- `putc_unlocked()`
- `putchar_unlocked()`
- `getc_unlocked()` (stub)
- `getchar_unlocked()` (stub)

ycsin committed:
posix: implement file locking functions

Implement file locking functions and added testcase.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
libc: minimal: implement putc_unlocked & putchar_unlocked

Add a simple implementation for putc_unlocked() &
putchar_unlocked().

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

cfriedt commented on lib/os/fdtable.c:
```diff
@@ -30,11 +30,20 @@ struct fd_entry {
 	const struct fd_op_vtable *vtable;
 	atomic_t refcount;
 	struct k_mutex lock;
+#if defined(CONFIG_POSIX_FILE_LOCKING)
```
I think we want to reuse the existing lock for this purpose.

ycsin committed:
libc: minimal: stubs getc_unlocked & getchar_unlocked

Add stub implementation and test for `getc_unlocked()` &
`getchar_unlocked()`.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

cfriedt commented on tests/posix/common/src/file_locking.c:
```diff
@@ -0,0 +1,148 @@
+#include <errno.h>
+#include <stdio.h>
+
+#ifdef CONFIG_PICOLIBC
+/* Not sure how to do this for picolibc */
+void flockfile(FILE *file);
+int ftrylockfile(FILE *file);
+void funlockfile(FILE *file);
+int putc_unlocked(int c, FILE *file);
+int putchar_unlocked(int c);
+int getc_unlocked(FILE *stream);
+int getchar_unlocked(void);
+#endif
```
If the declarations are missing from picolibc for whatever reason, I think this should be ok

ycsin replied:
Removed them for picolibc as there is no way for any application to use them

keith-packard replied:
Sounds like we just need to add these to picolibc. which standard are these from?

ycsin replied:
It's in the `_POSIX_THREAD_SAFE_FUNCTIONS` option group

https://pubs.opengroup.org/onlinepubs/9699919799/functions/flockfile.html

keith-packard replied:
https://github.com/picolibc/picolibc/pull/752

cfriedt reviewed (dismissed):
I think this is fine too for the time being. Thanks for adding tests

cfriedt commented on lib/posix/options/Kconfig.file_locking:
```diff
@@ -0,0 +1,37 @@
+
+config POSIX_FILE_LOCKING
+	bool "POSIX file locking [EXPERIMENTAL]"
+	select EXPERIMENTAL
```
This might require something like `FDTABLE` as well

cfriedt commented on lib/libc/common/source/stdio/getc_unlocked.c:
```diff
@@ -0,0 +1,24 @@
+
+#include <zephyr/sys/util.h>
+
+int getc_unlocked(FILE *stream)
+{
+	ARG_UNUSED(stream);
+
+	errno = ENOSYS;
+
+	return EOF;
+}
```
For this one, I would add a `zvfs_getc_unlocked()` and just call that.

`zvfs_getc_unlocked()` should do some basic error checking but should not incur lock overhead.

E.g. check to see if `stream` is a valid pointer with `IS_ARRAY_ELEMENT()`, check the `read()` vtable method is non-`NULL`, etc.
